lib/Analysis/Allocation.cpp

```diff
@@ -88,30 +88,74 @@ getScratchConfigForCvtLayout(triton::gpu::ConvertLayoutOp op, unsigned &inVec,
   return paddedRepShape;
 }
 
-SmallVector<unsigned> getScratchConfigForReduce(triton::ReduceOp op) {
+/// ====================== reduce op config begin ==========================
+
+bool enableFastReduce(triton::ReduceOp op) {
+  auto srcTy = op.operand().getType().cast<RankedTensorType>();
+  auto srcLayout = srcTy.getEncoding();
+  return op.axis() == getOrder(srcLayout)[0];
+}
+
+SmallVector<unsigned> getScratchConfigForReduceBasic(triton::ReduceOp op) {
   auto srcTy = op.operand().getType().cast<RankedTensorType>();
   auto srcLayout = srcTy.getEncoding();
   auto srcShape = srcTy.getShape();
   auto axis = op.axis();
 
-  bool fastReduce = axis == getOrder(srcLayout)[0];
+  auto smemShape = convertType<unsigned>(srcShape);
+  smemShape[axis] = gpu::getThreadsPerWarp(srcLayout)[axis] *
+                    gpu::getWarpsPerCTA(srcLayout)[axis];
 
-  SmallVector<unsigned> smemShape;
-  for (auto d : srcShape)
-    smemShape.push_back(d);
+  return smemShape;
+}
+
+SmallVector<SmallVector<unsigned>>
+getScratchConfigsForReduceFast(triton::ReduceOp op) {
+  auto srcTy = op.operand().getType().cast<RankedTensorType>();
+  auto srcLayout = srcTy.getEncoding();
+  auto srcShape = srcTy.getShape();
+  auto axis = op.axis();
+
+  SmallVector<SmallVector<unsigned>> smemShapes(3);
+
+  /// shared memory block0
```

Comment:
Asking a naive question, what is the exact meaning of block0/1/2 here?

Reply:
I may have made a mistake here. It seems that one shape is enough because the second and third shape is always smaller than the first one. Let me think about this.

Reply:
I leave a FIXME here and will fix this in another PR.